### ISSUE TITLE
Got "Stack level too deep" errors uploading files after upgrading to 6.15.0

### ISSUE THREAD

**Issue by henrik** [newrelic/newrelic-ruby-agent#548]
After upgrading from 6.14.0 to 6.15.0, Active Storage file uploads in a Rails 5.2.4.4 app (on Ruby 2.7.2) started failing.

The upload code is basically just

```
invoice.documents.attach(params[:invoice][:documents])
```

See below for logs (the time stamps will differ because the below was pieced together from separate runs).

We get the impression that New Relic and rack-mini-profiler get into an endless loop, causing the error. Downgrading New Relic fixed it.

```
2021-01-25T10:22:43.639186+00:00 app[web.1]:
2021-01-25T10:22:43.639348+00:00 app[web.1]: SystemStackError (stack level too deep):
2021-01-25T10:22:43.639371+00:00 app[web.1]:
2021-01-25T10:22:43.639455+00:00 app[web.1]: app/controllers/invoices_controller.rb:32:in `attach_documents'
…
2021-01-25T10:21:53.890829+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-2.3.0/lib/patches/net_patches.rb:9:in `block in request'
2021-01-25T10:21:53.890829+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-2.3.0/lib/mini_profiler/profiling_methods.rb:46:in `step'
2021-01-25T10:21:53.890829+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-2.3.0/lib/patches/net_patches.rb:8:in `request'
2021-01-25T10:21:53.890830+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/instrumentation/net.rb:52:in `block (2 levels) in request_with_newrelic_trace'
2021-01-25T10:21:53.890830+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/tracer.rb:371:in `capture_segment_error'
2021-01-25T10:21:53.890830+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/instrumentation/net.rb:51:in `block in request_with_newrelic_trace'
2021-01-25T10:21:53.890831+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent.rb:507:in `disable_all_tracing'
2021-01-25T10:21:53.890840+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/instrumentation/net.rb:50:in `request_with_newrelic_trace'
2021-01-25T10:21:53.890840+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-2.3.0/lib/patches/net_patches.rb:9:in `block in request'
2021-01-25T10:21:53.890841+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-2.3.0/lib/mini_profiler/profiling_methods.rb:46:in `step'
2021-01-25T10:21:53.890841+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-2.3.0/lib/patches/net_patches.rb:8:in `request'
2021-01-25T10:21:53.890841+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/instrumentation/net.rb:52:in `block (2 levels) in request_with_newrelic_trace'
2021-01-25T10:21:53.890842+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/tracer.rb:371:in `capture_segment_error'
2021-01-25T10:21:53.890842+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/instrumentation/net.rb:51:in `block in request_with_newrelic_trace'
2021-01-25T10:21:53.890850+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent.rb:507:in `disable_all_tracing'
2021-01-25T10:21:53.890850+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/instrumentation/net.rb:50:in `request_with_newrelic_trace'
…

**Comment by henrik** [comment]
Trying to repro in dev – haven't gotten the actual upload to fail in dev (yet?) but seeing this in the logs:

```
/home/henrik/.devbox/code/bookie/gems/gems/addressable-2.7.0/lib/addressable/uri.rb:463:in `initialize_dup': stack level too deep (SystemStackError)
#<Thread:0x000056350780d7d8 /home/henrik/.devbox/code/bookie/gems/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/threading/agent_thread.rb:12 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
        10377: from /home/henrik/.devbox/code/bookie/gems/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/threading/agent_thread.rb:14:in `block in create'
        10376: from /home/henrik/.devbox/code/bookie/gems/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/agent.rb:723:in `block in start_worker_thread'
        10375: from /home/henrik/.devbox/code/bookie/gems/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/agent.rb:696:in `deferred_work!'
        10374: from /home/henrik/.devbox/code/bookie/gems/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/agent.rb:677:in `catch_errors'
        10373: from /home/henrik/.devbox/code/bookie/gems/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/agent.rb:697:in `block in deferred_work!'
        10372: from /home/henrik/.devbox/code/bookie/gems/gems/newrelic_rpm-6.15.0/lib/new_relic/agent.rb:507:in `disable_all_tracing'
        10371: from /home/henrik/.devbox/code/bookie/gems/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/agent.rb:698:in `block (2 levels) in deferred_work!'
        10370: from /home/henrik/.devbox/code/bookie/gems/gems/newrelic_rpm-6.15.0/lib/new_relic/agent/agent.rb:943:in `connect'
         ... 10365 levels...
            4: from /home/henrik/.devbox/code/bookie/gems/gems/addressable-2.7.0/lib/addressable/uri.rb:831:in `block in initialize'
            3: from /home/henrik/.devbox/code/bookie/gems/gems/addressable-2.7.0/lib/addressable/uri.rb:1401:in `port='
            2: from /home/henrik/.devbox/code/bookie/gems/gems/addressable-2.7.0/lib/addressable/uri.rb:463:in `unencode'
            1: from /home/henrik/.devbox/code/bookie/gems/gems/addressable-2.7.0/lib/addressable/uri.rb:463:in `dup'
```

**Comment by tannalynn** [comment]
Hello @henrik. Taking a look at this, it appears to be related to the compatibility issues between Module.prepend and monkeypatching. You pointed out that removing that require on rack-mini-profiler fixes the problem, and one of those is `prepend_net_http_patch`. This is what first caught my attention as being related to the prepend VS monkeypatching situation. 

So, before  6.14, our Net::HTTP instrumentation used monkeypatching exclusively. 
In 6.14, we introduced another way to instrument Net::HTTP, using Module.prepend. We added a config option `prepend_net_instrumentation` to control this behavior so users can decide which style of instrumentation to use. Additionally, we also set the default for this option to true, using prepend by default to instrument Net::HTTP. 
Using prepend by default caused a number of users to experience errors due to various other gems using monkeypatching. Changing the default behavior of the agent in a way that caused unexpected errors is something we decided should be saved for our next major release instead. So for our 6.15 release, we changed this default to false, so it now uses monkeypatching by default again. 

The option to use prepend is still available by adding the following config option to your newrelic.yml 
`prepend_net_instrumentation: true`
I believe adding this should solve this error for you.

Hopefully this helps to explain the differences here in 6.14. Please let me know if you run into any issues or have any questions.

**Comment by angelatan2** [comment]
Hello @henrik, Thanks again for filling the issue.  I'm checking in if you have considered the suggestion from @tannalynn to set prepend_net_instrumentation: true.  If you did,  could you share if it resolves the problem that you saw?

**Comment by henrik** [comment]
Thank you for that thorough response, and apologies for not getting back to you sooner.

We ended up just changing

    gem "rack-mini-profiler", require: [ "prepend_net_http_patch", "rack-mini-profiler" ]

to

    gem "rack-mini-profiler"

which made this error go away. Haven't noticed any issues since with either lib, so we're happy with that :) 

I'll close this issue (assuming you don't mind) but if anyone else runs into it and finds their way here, and can't make the change we did, `prepend_net_instrumentation: true` seems like a great thing to try.
